File: src/data/practicalProgress.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import {
  EMPTY_PRACTICAL_PROGRESS, recordCaseStep, recordLabAnswered, recordStationRun,
  setSkillStatus, summariseSkills,
} from './practicalProgress.ts'

const AT = '2026-08-13T10:00:00.000Z'
const LATER = '2026-08-14T10:00:00.000Z'

test('a station run counts an attempt and sets the first best score', () => {
  const after = recordStationRun(EMPTY_PRACTICAL_PROGRESS, 'os-cvs', { marks: 14, outOf: 20, checkedItems: ['a'], at: AT })
  assert.deepEqual(after.stations['os-cvs'], {
    attempts: 1, bestMarks: 14, outOf: 20, lastAt: AT, checkedItems: ['a'],
  })
})

test('a worse run still counts as an attempt but does not lower the best', () => {
  const first = recordStationRun(EMPTY_PRACTICAL_PROGRESS, 'os-cvs', { marks: 18, outOf: 20, checkedItems: [], at: AT })
  const second = recordStationRun(first, 'os-cvs', { marks: 9, outOf: 20, checkedItems: ['x'], at: LATER })
  assert.equal(second.stations['os-cvs'].attempts, 2)
  assert.equal(second.stations['os-cvs'].bestMarks, 18)
  // The resumable ticks are always the most recent run, not the best one.
  assert.deepEqual(second.stations['os-cvs'].checkedItems, ['x'])
})

test('the best score carries the mark total it was scored against', () => {
  // 18/20 is better than 18/30; comparing raw marks would call them equal and
  // let a re-authored station rewrite an old result's meaning.
  const first = recordStationRun(EMPTY_PRACTICAL_PROGRESS, 's', { marks: 18, outOf: 20, checkedItems: [], at: AT })
  const second = recordStationRun(first, 's', { marks: 18, outOf: 30, checkedItems: [], at: LATER })
  assert.equal(second.stations['s'].bestMarks, 18)
  assert.equal(second.stations['s'].outOf, 20)
})

test('a station scored out of nothing does not divide by zero', () => {
  const after = recordStationRun(EMPTY_PRACTICAL_PROGRESS, 's', { marks: 0, outOf: 0, checkedItems: [], at: AT })
  assert.equal(after.stations['s'].attempts, 1)
  assert.equal(after.stations['s'].bestMarks, 0)
})

test('a case in progress becomes completed and stays completed', () => {
  const started = recordCaseStep(EMPTY_PRACTICAL_PROGRESS, 'c1', { lastStep: 2, steps: 6, completed: false, at: AT })
  assert.equal(started.cases['c1'].status, 'in-progress')
  const finished = recordCaseStep(started, 'c1', { lastStep: 6, steps: 6, completed: true, at: LATER })
  assert.equal(finished.cases['c1'].status, 'completed')
  // Reopening a finished case to reread it must not undo having finished it.
  const revisited = recordCaseStep(finished, 'c1', { lastStep: 1, steps: 6, completed: false, at: LATER })
  assert.equal(revisited.cases['c1'].status, 'completed')
  assert.equal(revisited.cases['c1'].lastStep, 6)
})

test('lab progress never goes backwards on a revisit', () => {
  const first = recordLabAnswered(EMPTY_PRACTICAL_PROGRESS, 'l1', { done: 8, items: 20, at: AT })
  const second = recordLabAnswered(first, 'l1', { done: 3, items: 20, at: LATER })
  assert.equal(second.labs['l1'].done, 8)
  assert.equal(second.labs['l1'].lastAt, LATER)
})

test('a skill can be practised, then marked ready', () => {
  const practised = setSkillStatus(EMPTY_PRACTICAL_PROGRESS, 'sk1', 'practised', AT)
  assert.equal(practised.skills['sk1'].status, 'practised')
  const ready = setSkillStatus(practised, 'sk1', 'ready', LATER)
  assert.equal(ready.skills['sk1'].status, 'ready')
})

test('clearing a skill removes it rather than storing a blank state', () => {
  const practised = setSkillStatus(EMPTY_PRACTICAL_PROGRESS, 'sk1', 'practised', AT)
  const cleared = setSkillStatus(practised, 'sk1', 'not-started', LATER)
  assert.equal('sk1' in cleared.skills, false)
})

test('the skills headline is counted from the list it is shown above', () => {
  // The old headline said "14 / 22" over a list of twelve. The total is passed
  // in from the live list, so the two cannot disagree.
  let progress = setSkillStatus(EMPTY_PRACTICAL_PROGRESS, 'a', 'practised', AT)
  progress = setSkillStatus(progress, 'b', 'ready', AT)
  progress = setSkillStatus(progress, 'c', 'ready', AT)
  assert.deepEqual(summariseSkills(progress, 12), { practised: 3, ready: 2, total: 12 })
})

test('an untouched record summarises as nothing done', () => {
  assert.deepEqual(summariseSkills(EMPTY_PRACTICAL_PROGRESS, 12), { practised: 0, ready: 0, total: 12 })
})

File: android/app/src/test/java/com/synapse/android/core/practical/PracticalProgressTest.kt
package com.synapse.android.core.practical

import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * [PracticalProgress] and its fold functions, ported from
 * `src/data/practicalProgress.ts`. Every test here works against real
 * encode/decode through [json], not just in-memory data-class copying --
 * that is what [`a section this app never touches survives a write`] is
 * actually checking: not that Kotlin can copy a field, but that the decoder
 * this app ships does not silently drop one.
 */
class PracticalProgressTest {

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * A `synapse.practical.progress.v1` document as the *web* leaves it --
     * not as Kotlin would encode it. That distinction is the whole test:
     * anything built from [PracticalProgress] can only contain fields Kotlin
     * declares, so round-tripping one proves nothing about the documents
     * this app will actually be handed.
     *
     * `lastScorePct` and `updatedAt` below are real-shaped fields Kotlin does
     * not declare.
     */
    private val webWritten = """
        {
          "version": 1,
          "updatedAt": "2026-08-01T09:12:44.000Z",
          "stations": {},
          "cases": {
            "cc-chest": {"status":"in-progress","lastStep":2,"steps":5,"lastAt":"2026-08-01T00:00:00Z"}
          },
          "labs": {
            "li-abg": {"done":4,"items":20,"lastAt":"2026-08-01T00:00:00Z","lastScorePct":80}
          },
          "skills": {
            "sk-bp": {"status":"ready","lastAt":"2026-08-01T00:00:00Z"}
          }
        }
    """.trimIndent()

    @Test
    fun `a section this app never touches survives a write`() {
        val decoded = json.decodeFromString(PracticalProgress.serializer(), webWritten)

        // Simulate an Android write that only ever touches stations.
        val afterStationWrite = recordStationRun(decoded, "os-cvs", marks = 10, outOf = 24, checkedItems = listOf("m1"), at = "2026-08-02T00:00:00Z")
        val reEncoded = json.encodeToString(PracticalProgress.serializer(), afterStationWrite)
        val final = json.decodeFromString(PracticalProgress.serializer(), reEncoded)

        assertEquals(
            CaseProgress(status = "in-progress", lastStep = 2, steps = 5, lastAt = "2026-08-01T00:00:00Z"),
            final.cases.getValue("cc-chest"),
        )
        assertEquals(LabProgress(done = 4, items = 20, lastAt = "2026-08-01T00:00:00Z"), final.labs.getValue("li-abg"))
        assertEquals(SkillProgress(status = "ready", lastAt = "2026-08-01T00:00:00Z"), final.skills.getValue("sk-bp"))
        assertEquals(10, final.stations.getValue("os-cvs").bestMarks)
    }

    @Test
    fun `what an Android write does not preserve is the document growing`() {
        // The honest limit of ignoreUnknownKeys, pinned so the KDoc on
        // PracticalProgress cannot drift back to overclaiming. A key Kotlin
        // does not declare gets through the decoder and is then dropped,
        // because nothing carries it to the encoder -- and a whole new
        // top-level section a later client adds is such a key. Making this
        // test fail is the point at which Android is ready for one.
        val withNewSection = webWritten.replace(
            """  "skills": {""",
            """  "vivas": {"vv-cvs": {"status":"ready","lastAt":"2026-08-01T00:00:00Z"}},
  "skills": {""",
        )
        val decoded = json.decodeFromString(PracticalProgress.serializer(), withNewSection)
        val reEncoded = json.encodeToString(
            PracticalProgress.serializer(),
            recordStationRun(decoded, "os-cvs", marks = 10, outOf = 24, checkedItems = emptyList(), at = "t1"),
        )

        assertTrue(withNewSection.contains("vivas"))
        assertFalse(reEncoded.contains("vivas"))
        assertFalse(reEncoded.contains("lastScorePct"))
        assertFalse(reEncoded.contains("updatedAt"))
    }

    @Test
    fun `the best score only moves up`() {
        var progress = PracticalProgress()
        progress = recordStationRun(progress, "os-cvs", marks = 20, outOf = 24, checkedItems = emptyList(), at = "t1")
        progress = recordStationRun(progress, "os-cvs", marks = 10, outOf = 24, checkedItems = emptyList(), at = "t2")

        assertEquals(20, progress.stations.getValue("os-cvs").bestMarks)
    }

    @Test
    fun `a worse run still updates the ticks to resume from`() {
        var progress = PracticalProgress()
        progress = recordStationRun(progress, "os-cvs", marks = 20, outOf = 24, checkedItems = listOf("m1", "m2"), at = "t1")
        progress = recordStationRun(progress, "os-cvs", marks = 5, outOf = 24, checkedItems = listOf("m3"), at = "t2")

        val station = progress.stations.getValue("os-cvs")
        assertEquals(20, station.bestMarks)
        assertEquals(listOf("m3"), station.checkedItems)
    }

    @Test
    fun `a best score carries the total it was scored against`() {
        var progress = PracticalProgress()
        progress = recordStationRun(progress, "os-cvs", marks = 18, outOf = 20, checkedItems = emptyList(), at = "t1")
        // Re-authored to 30 marks; a worse share must not overwrite the total.
        progress = recordStationRun(progress, "os-cvs", marks = 20, outOf = 30, checkedItems = emptyList(), at = "t2")

        val station = progress.stations.getValue("os-cvs")
        assertEquals(18, station.bestMarks)
        assertEquals(20, station.outOf)
    }

    @Test
    fun `an equal share is the better run, so the newer mark total wins`() {
        var progress = PracticalProgress()
        progress = recordStationRun(progress, "os-cvs", marks = 9, outOf = 10, checkedItems = emptyList(), at = "t1")
        // Re-authored out of 20, sat again, same 90% share. `>=`, not `>`:
        // the student's record should follow the station they actually sat,
        // and "18 / 20" is the honest way to show today's 90%.
        progress = recordStationRun(progress, "os-cvs", marks = 18, outOf = 20, checkedItems = emptyList(), at = "t2")

        val station = progress.stations.getValue("os-cvs")
        assertEquals(18, station.bestMarks)
        assertEquals(20, station.outOf)
        assertEquals(2, station.attempts)
    }

    @Test
    fun `a station scored out of zero does not divide by zero`() {
        val progress = recordStationRun(PracticalProgress(), "os-cvs", marks = 0, outOf = 0, checkedItems = emptyList(), at = "t1")

        val station = progress.stations.getValue("os-cvs")
        assertEquals(0, station.bestMarks)
        assertEquals(0, station.outOf)
        assertEquals(1, station.attempts)
    }

    @Test
    fun `a completed case is not demoted by revisiting it`() {
        var progress = PracticalProgress()
        progress = recordCaseStep(progress, "cc-chest", lastStep = 5, steps = 5, completed = true, at = "t1")
        progress = recordCaseStep(progress, "cc-chest", lastStep = 0, steps = 5, completed = false, at = "t2")

        assertEquals("completed", progress.cases.getValue("cc-chest").status)
    }

    @Test
    fun `lastStep never goes backwards`() {
        var progress = PracticalProgress()
        progress = recordCaseStep(progress, "cc-chest", lastStep = 3, steps = 5, completed = false, at = "t1")
        progress = recordCaseStep(progress, "cc-chest", lastStep = 1, steps = 5, completed = false, at = "t2")

        assertEquals(3, progress.cases.getValue("cc-chest").lastStep)
    }

    @Test
    fun `a lab's done count never goes backwards`() {
        var progress = PracticalProgress()
        progress = recordLabAnswered(progress, "li-abg", done = 12, items = 20, at = "t1")
        progress = recordLabAnswered(progress, "li-abg", done = 4, items = 20, at = "t2")

        assertEquals(12, progress.labs.getValue("li-abg").done)
    }

    @Test
    fun `setting a skill back to not-started removes it`() {
        var progress = PracticalProgress()
        progress = setSkillStatus(progress, "sk-bp", "practised", at = "t1")
        assertTrue(progress.skills.containsKey("sk-bp"))

        progress = setSkillStatus(progress, "sk-bp", "not-started", at = "t2")

        assertFalse(progress.skills.containsKey("sk-bp"))
    }

    @Test
    fun `practised counts the ready ones too`() {
        var progress = PracticalProgress()
        progress = setSkillStatus(progress, "sk-bp", "practised", at = "t1")
        progress = setSkillStatus(progress, "sk-cvs", "ready", at = "t1")
        progress = setSkillStatus(progress, "sk-resp", "not-started", at = "t1")

        val summary = summariseSkills(progress, total = 12)

        assertEquals(2, summary.practised)
        assertEquals(1, summary.ready)
        assertEquals(12, summary.total)
    }
}
